pkg/backup/backupdest/backup_index.go

```diff
@@ -123,6 +126,172 @@ func IndexExists(ctx context.Context, store cloud.ExternalStorage, subdir string
+		if times, ok := timeMemo[basename]; ok {
+			return times[0], times[1], nil
+		}
+		start, end, err := parseIndexFilename(basename)
+		if err != nil {
+			return time.Time{}, time.Time{}, err
+		}
+		timeMemo[basename] = [2]time.Time{start, end}
+		return start, end, nil
+	}
+	var sortErr error
+	slices.SortFunc(indexBasenames, func(a, b string) int {
+		aStart, aEnd, err := indexTimesFromFile(a)
+		if err != nil {
+			sortErr = err
+		}
+		bStart, bEnd, err := indexTimesFromFile(b)
+		if err != nil {
+			sortErr = err
+		}
+		if aEnd.Before(bEnd) {
+			return -1
+		} else if aEnd.After(bEnd) {
+			return 1
+		}
+		// End times are equal, so break tie with start time.
+		if aStart.Before(bStart) {
+			return -1
+		} else {
+			return 1
+		}
+	})
+	if sortErr != nil {
+		return nil, errors.Wrapf(sortErr, "sorting index filenames")
+	}
+
+	return indexBasenames, nil
+}
+
+// GetBackupTreeIndexMetadata concurrently retrieves the index metadata for all
+// backups within the specified subdir, up to the specified end time, inclusive.
+// The store should be rooted at the collection URI that contains the `index/`
+// directory. Indexes are returned in ascending end time order, with ties broken
+// by ascending start time order. If the end time is not covered by the backups
+// in the subdir, an error is returned.
+//
+// Note: If endTime is provided, GetBackupTreeIndexMetadata will return ALL
+// backups that could be used to restore to endTime. So even if a compacted
+// backup can be used to restore to endTime, the incremental backups that
+// make up the compacted backup will also be returned.
+func GetBackupTreeIndexMetadata(
+	ctx context.Context, store cloud.ExternalStorage, subdir string, endTime hlc.Timestamp,
+) ([]backuppb.BackupIndexMetadata, error) {
+	indexBasenames, err := ListIndexes(ctx, store, subdir)
+	if err != nil {
+		return nil, err
+	}
+
+	indexes := make([]backuppb.BackupIndexMetadata, len(indexBasenames))
+	g := ctxgroup.WithContext(ctx)
+	for i, basename := range indexBasenames {
+		g.GoCtx(func(ctx context.Context) error {
+			reader, size, err := store.ReadFile(
+				ctx, path.Join(indexSubdir(subdir), basename), cloud.ReadOptions{},
+			)
+			if err != nil {
+				return errors.Wrapf(err, "reading index file %s", basename)
+			}
+			defer reader.Close(ctx)
+
+			bytes := make([]byte, size)
+			if _, err := reader.Read(ctx, bytes); err != nil {
+				return errors.Wrapf(err, "reading index file %s bytes", basename)
+			}
+
+			index := backuppb.BackupIndexMetadata{}
+			if err := protoutil.Unmarshal(bytes, &index); err != nil {
+				return errors.Wrapf(err, "unmarshalling index file %s", basename)
+			}
+			indexes[i] = index
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return nil, errors.Wrapf(err, "getting backup index metadata")
+	}
+
+	if endTime.IsEmpty() {
+		return indexes, nil
+	}
+
+	coveringIdx := slices.IndexFunc(indexes, func(index backuppb.BackupIndexMetadata) bool {
```

Comment:
i've realized that we can probably remove this truncation logic, as ValidateEndTimeAndTruncate is gonna do it anyway. But, we could keep it in for now until the downstream PRs land.